### ISSUE TITLE
Instrumentation for StrictMode

### ISSUE THREAD

**Issue by MrHadiSatrio** [open-telemetry/opentelemetry-android#520]
Hello. Are there any plans for StrictMode instrumentation? `ThreadPolicy#penaltyListener()` and `VmPolicy#penaltyListener()` which were added in API 28 may provide a solid foundation for one.

**Comment by LikeTheSalad** [comment]
That sounds interesting, though it seems like StrictMode is more targeted as a development tool, so I'm wondering if there's a non-trivial performance overhead caused by enabling it?

**Comment by LikeTheSalad** [comment]
Thank you for taking the time to collect and provide this info 🙏 it helps a lot. Based on what you found I'd say that we shouldn't provide this kind of instrumentation by default to avoid affecting apps' performance unless it's really needed. It's still probably worth creating some opt-in instrumentation for it, though I'd like to get some people to chime in just to get an idea of what's the appetite for it.

**Comment by MrHadiSatrio** [comment]
I agree with you. One thing I want to also highlight is that the above example essentially only triggers one type of violation. So in an actual application, there could be an even bigger overhead.

That said, in my honest opinion, there is still value to be gained, especially (or perhaps exclusively) in staging environments. `penaltyDeath()` is not always an option, so the logs may enable teams to stay on top of these violations and take action.

Curious to hear what others think as well! 🙌🏽

**Comment by breedx-splk** [comment]
@MrHadiSatrio this is interesting for sure. Is the expected usage for application developers to have additional ways of leveraging `StrictMode` through this otel agent, versus wiring it in themselves? And then the idea is to export those violations as logs when they occur?

I think it's perfectly acceptable for us to have instrumentation in this repo that isn't necessarily enabled by default or included in the topmost agent. Seems like a worthy experiment.

Is it also possible to turn these strict mode configurations back off?

**Comment by MrHadiSatrio** [comment]
So I took a quick stab at the idea and did a spike implementation [on this branch](https://github.com/MrHadiSatrio/opentelemetry-android/tree/feature/strictmode_instrumentation). I couldn't seem to surface the logs on Jaeger, so I went with [grafana/otel-lgtm](https://hub.docker.com/r/grafana/otel-lgtm) instead.

Here's what a StrictMode dashboard may look like:

<img width="1624" alt="Screenshot 2024-08-15 at 14 11 59" src="https://github.com/user-attachments/assets/feaec82f-afdf-4999-964f-fa1a3f3b4586">

**Comment by bidetofevil** [comment]
I've found that strict mode violation tend to be the result of structural issues with the code, so deploying it on a beta or dogfood build is usually sufficient when trying to ferret out said issues. If we were add this to the project, I think making this opt-in would be desirable so that it can be enabled for certain flavors or builds, but not in general production.
